File: Investment_App/BackendFastAPI/app/api/v1/portfolio.py
"""
app/api/v1/portfolio.py
───────────────────────
Portfolio endpoints:
  GET /api/v1/{broker}/holdings
  GET /api/v1/{broker}/positions
  GET /api/v1/{broker}/trades
  GET /api/v1/{broker}/summary
  POST /api/v1/{broker}/cache/invalidate
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Path

from app.api.deps import get_portfolio_service
from app.core.exceptions import BrokerAuthError, BrokerError, BrokerNotFoundError
from app.models.portfolio import Holding, PortfolioSummary, Position, Trade
from app.models.responses import APIResponse
from app.services.portfolio_service import PortfolioService

router = APIRouter(tags=["portfolio"])

_ServiceDep = Annotated[PortfolioService, Depends(get_portfolio_service)]


def _resolve(broker_id: str, svc: PortfolioService):
    """Validate broker and raise proper HTTP errors."""
    try:
        svc._require(broker_id)
    except BrokerNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/{broker}/holdings", response_model=APIResponse[list[Holding]])
def get_holdings(
    broker: Annotated[str, Path(description="Broker identifier, e.g. 'upstox'")],
    svc: _ServiceDep,
) -> APIResponse[list[Holding]]:
    _resolve(broker, svc)
    try:
        return APIResponse(data=svc.get_holdings(broker))
    except BrokerAuthError as exc:
        raise HTTPException(status_code=401, detail=str(exc)) from exc
    except BrokerError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.get("/{broker}/positions", response_model=APIResponse[list[Position]])
def get_positions(
    broker: Annotated[str, Path()],
    svc: _ServiceDep,
) -> APIResponse[list[Position]]:
    _resolve(broker, svc)
    try:
        return APIResponse(data=svc.get_positions(broker))
    except BrokerAuthError as exc:
        raise HTTPException(status_code=401, detail=str(exc)) from exc
    except BrokerError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.get("/{broker}/trades", response_model=APIResponse[list[Trade]])
def get_trades(
    broker: Annotated[str, Path()],
    svc: _ServiceDep,
) -> APIResponse[list[Trade]]:
    _resolve(broker, svc)
    try:
        return APIResponse(data=svc.get_trades(broker))
    except BrokerAuthError as exc:
        raise HTTPException(status_code=401, detail=str(exc)) from exc
    except BrokerError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.get("/{broker}/summary", response_model=APIResponse[PortfolioSummary])
def get_summary(
    broker: Annotated[str, Path()],
    svc: _ServiceDep,
) -> APIResponse[PortfolioSummary]:
    _resolve(broker, svc)
    try:
        return APIResponse(data=svc.get_summary(broker))
    except BrokerAuthError as exc:
        raise HTTPException(status_code=401, detail=str(exc)) from exc
    except BrokerError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.post("/{broker}/cache/invalidate", response_model=APIResponse[dict])
def invalidate_cache(
    broker: Annotated[str, Path()],
    svc: _ServiceDep,
) -> APIResponse[dict]:
    _resolve(broker, svc)
    svc.invalidate(broker)
    return APIResponse(data={"message": f"Cache cleared for broker '{broker}'"})

File: Investment_App/TradeView/src/api/portfolio.ts
import api from './index'
import type {
  ApiEnvelope,
  Holding,
  Position,
  Trade,
  PortfolioSummary,
  AnalysisResult,
  Alert,
  BrokerInfo,
} from '@/types/portfolio'

// ─── Portfolio API ────────────────────────────────────────────────────────────
// Mirrors ClientConsolApp/Services/PortfolioService.cs — each method maps 1:1
// to a FastAPI endpoint under /api/v1/{broker}/...

export const portfolioApi = {
  /** GET /api/v1/brokers — list all registered broker adapters (etoro, upstox) */
  async getBrokers(): Promise<BrokerInfo[]> {
    const { data } = await api.get<ApiEnvelope<BrokerInfo[]>>('/api/v1/brokers')
    return data.data
  },

  /** GET /api/v1/{broker}/holdings */
  async getHoldings(broker: string): Promise<Holding[]> {
    const { data } = await api.get<ApiEnvelope<Holding[]>>(`/api/v1/${broker}/holdings`)
    return data.data
  },

  /** GET /api/v1/{broker}/positions */
  async getPositions(broker: string): Promise<Position[]> {
    const { data } = await api.get<ApiEnvelope<Position[]>>(`/api/v1/${broker}/positions`)
    return data.data
  },

  /** GET /api/v1/{broker}/trades */
  async getTrades(broker: string): Promise<Trade[]> {
    const { data } = await api.get<ApiEnvelope<Trade[]>>(`/api/v1/${broker}/trades`)
    return data.data
  },

  /** GET /api/v1/{broker}/summary */
  async getSummary(broker: string): Promise<PortfolioSummary> {
    const { data } = await api.get<ApiEnvelope<PortfolioSummary>>(`/api/v1/${broker}/summary`)
    return data.data
  },

  /** GET /api/v1/{broker}/analysis */
  async getAnalysis(broker: string): Promise<AnalysisResult> {
    const { data } = await api.get<ApiEnvelope<AnalysisResult>>(`/api/v1/${broker}/analysis`)
    return data.data
  },

  /** GET /api/v1/{broker}/analysis/alerts */
  async getAlerts(broker: string): Promise<Alert[]> {
    const { data } = await api.get<ApiEnvelope<Alert[]>>(`/api/v1/${broker}/analysis/alerts`)
    return data.data
  },

  /** POST /api/v1/{broker}/cache/invalidate — force fresh data fetch from broker */
  async invalidateCache(broker: string): Promise<{ message: string }> {
    const { data } = await api.post<ApiEnvelope<{ message: string }>>(
      `/api/v1/${broker}/cache/invalidate`,
    )
    return data.data
  },
}
